ibis/backends/base_sql/__init__.py

```diff
@@ -125,3 +136,1013 @@ def _name_expr(formatted_expr, quoted_name):
+    arg_formatted = translator.translate(arg)
+    start_formatted = translator.translate(start)
+
+    # Impala is 1-indexed
+    if length is None or isinstance(length.op(), ops.Literal):
+        lvalue = length.op().value if length is not None else None
+        if lvalue:
+            return 'substr({}, {} + 1, {})'.format(
+                arg_formatted, start_formatted, lvalue
+            )
+        else:
+            return 'substr({}, {} + 1)'.format(arg_formatted, start_formatted)
+    else:
+        length_formatted = translator.translate(length)
+        return 'substr({}, {} + 1, {})'.format(
+            arg_formatted, start_formatted, length_formatted
+        )
+
+
+def _string_find(translator, expr):
+    op = expr.op()
+    arg, substr, start, _ = op.args
+    arg_formatted = translator.translate(arg)
+    substr_formatted = translator.translate(substr)
+
+    if start is not None and not isinstance(start.op(), ops.Literal):
+        start_fmt = translator.translate(start)
+        return 'locate({}, {}, {} + 1) - 1'.format(
+            substr_formatted, arg_formatted, start_fmt
+        )
+    elif start is not None and start.op().value:
+        sval = start.op().value
+        return 'locate({}, {}, {}) - 1'.format(
+            substr_formatted, arg_formatted, sval + 1
+        )
+    else:
+        return 'locate({}, {}) - 1'.format(substr_formatted, arg_formatted)
+
+
+def _string_join(translator, expr):
+    op = expr.op()
+    arg, strings = op.args
+    return _format_call(translator, 'concat_ws', arg, *strings)
+
+
+def _extract_field(sql_attr):
+    def extract_field_formatter(translator, expr):
+        op = expr.op()
+        arg = translator.translate(op.args[0])
+
+        # This is pre-2.0 Impala-style, which did not used to support the
+        # SQL-99 format extract($FIELD from expr)
+        return "extract({}, '{}')".format(arg, sql_attr)
+
+    return extract_field_formatter
+
+
+def _find_in_set(translator, expr):
+    op = expr.op()
+
+    arg, str_list = op.args
+    arg_formatted = translator.translate(arg)
+    str_formatted = ','.join([x._arg.value for x in str_list])
+    return "find_in_set({}, '{}') - 1".format(arg_formatted, str_formatted)
+
+
+def _string_like(translator, expr):
+    arg, pattern, _ = expr.op().args
+    return '{} LIKE {}'.format(
+        translator.translate(arg), translator.translate(pattern)
+    )
+
+
+def _parse_url(translator, expr):
+    op = expr.op()
+
+    arg, extract, key = op.args
+    arg_formatted = translator.translate(arg)
+
+    if key is None:
+        return "parse_url({}, '{}')".format(arg_formatted, extract)
+    else:
+        key_fmt = translator.translate(key)
+        return "parse_url({}, '{}', {})".format(
+            arg_formatted, extract, key_fmt
+        )
+
+
+def _extract_epoch_seconds(t, expr):
+    (arg,) = expr.op().args
+    return 'unix_timestamp({})'.format(t.translate(arg))
+
+
+_impala_unit_names = {
```

Comment:
This, and the next method, seem to be specific to Impala, we shouldn't move it to the base sql backend.

Reply:
Thanks @datapythonista 

tying back to my below comment to @jreback - will there be separate PRs for moving each backend to `base_sql` (i.e. the focus of this PR is correctly setting it up and thats it)

If it is the case I would think items 2-4 from my list above would be handled in separate PRs.

What do you think?

Reply:
separate PRs are prob best, though also happy to move a 'bunch of stuff' then clean up later (it is slightly harder to remove stuff this way); whichever way easier / makes more clear what is changing.

another way to approach is to move a complete piece of functionaility, and from all backends at once, that way moving whole things and PRs are reasonable and clear what is happening.

Reply:
Moving some functions at a time in different PRs sounds good. I think in this case may make things trickier, but if you leave the registry itself to the last PR can make sense.

But just to be clear. Updating one backend at a time (one PR for the imports of bigquery, another for clickhouse) means that we need to duplicate the functions in the Impala backend and in base_sql. I think we should avoid that. And the imports are very small, moving the code is the big part.

So, small PRs is fine, if splitted by different operators, but better let's not do one backend at a time.

Reply:
ok, makes sense. Thanks @jreback @datapythonista 

So if i were to break it up by function type would it be good to do a separate PR per group within the operation_registry (although im not sure if this is contradictory to @datapythonista point of leaving registry itself to the last PR)?

The groups are (as identified by comments):

- Unary operations
- Unary aggregates
- String operations
- Timestamp operations
- Other operations
- RowNumber, and rank functions starts with 0 in Ibis-land

Still learning the structure of ibis so any guidance on approach is much appreciated!

Thx.

Reply:
I mentioned leaving the registry to the end so you don't need to import impala from base sql, which will create circular dependencies.

But feel free to do it anyway you feel like, I'd just avoid havibg the same function twice, which will create a mess.

Starting by the beginning of the registry, move the first 10 functions (or any number), open a PR, and then continue with the next 10 sounds like a reasonable thing to do. But up to you, I'd move stuff, and once you're in a consistent state and you want a break, just open a PR for what you've got.